core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java

```diff
@@ -84,13 +75,24 @@ protected void requestRefresh() {
   }
 
   protected String writeNewMetadata(TableMetadata metadata, int version) {
-    String newTableMetadataFilePath = newTableMetadataFilePath(metadata, version);
-    OutputFile newMetadataLocation = fileIo.newOutputFile(newTableMetadataFilePath);
+    FileIO initializedFileIo = io();
+    // This would be false if the subclass overrides io() and returns a different kind.
+    if (initializedFileIo instanceof HadoopFileIO) {
+      ((HadoopFileIO) initializedFileIo).updateTableMetadata(metadata);
```

Comment:
Alternatively, we could pass TableMetadata to `newMetadataOutputFile` and `newDataOutputFile`. Then we wouldn't need this notification.

Reply:
The problem is that `TableMetadata` is in `iceberg-core` but this module is in `iceberg-api`. But `FileIO` is returned by `Table`, which is in `iceberg-api`. Thus to make this API take `TableMetadata` as input we'd likely have to move `TableMetadata` into `iceberg-api`. The modules aren't set up correctly right now.

Reply:
Moving `TableMetadata` into `iceberg-api` runs into the problem of `TableOperations` not being in iceberg-api.

Reply:
I think better would be that instead of caching the `FileIO` object and trying to maintain its state, we should just make a new one with the most up to date metadata every time it's needed to write something.

Reply:
> I think we would expose the classes themselves. We'd add an interface to the API and implement that. That gets around the TableOperations problem.

Meaning, add an interface on top of `TableMetadata`?